### DISCUSSION TITLE
Job Fails When Using Custom Container Image

### DISCUSSION THREAD

**Discussion by lots0logs** [community/community#25070] [Actions]
<p>Following <a href="https://github.community/t/using-matrix-variable-in-docker-image-name/17296/4">this</a> advice I configured a custom container image for my build job. After doing so, the workflow fails at a step that works fine when no custom container is defined. I’m not sure what the problem is but my guess is possibly something to do with the workspace directory mount inside the container?</p>
<p>The step named “Init Submodules &amp; Checkout Branches” works fine with this config:</p>
<pre><code>name: CI
on: [push]
env:
  REPO: ${{ github.event.repository.name }}

jobs:
  #=========#
  # 1. Lint #
  #=========#
  lint:
    if: "! contains(toJSON(github.event.commits.*.msg), '[skip-ci]')"
    runs-on: ubuntu-18.04
    strategy:
      matrix:
        lang: ['php:7.4.1', 'php:5.6.40', 'php:5.2.17', 'js:12.14.0']

    steps:
      #======================#
      # 1.1 Initialize Repos #
      #======================#
      - name: Maybe Change REPO Environment Variable
        if: contains(github.event.repository.name, 'submodule')
        shell: bash
        run: echo "::set-env name=REPO::Divi"

      - name: Checkout Meta Repo
        uses: actions/checkout@v2
        with:
          repository: elegantthemes/meta
          ref: master
          path: workspace/meta
          token: ${{ secrets.ETSTAGING_TOKEN }}

      - name: Init Submodules &amp; Checkout Branches
        working-directory: workspace/meta
        shell: bash
        run: |
          git config --global http.extraheader "$(git config --get http.https://github.com/.extraheader)"
          git config --global protocol.version 2
          git submodule update --init --force --depth=1 --remote core-products
          cd core-products
          git submodule update --init --force --depth=1 --remote --recursive ${REPO}
          cd "${REPO}"
          git checkout "${GITHUB_REF}" || true
          git submodule foreach --recursive "git -c \"http.extraheader=$auth\" -c protocol.version=2 checkout \"${GITHUB_REF}\" || true"
</code></pre>
<p>It fails when I define a container though:</p>
<pre><code>name: CI
on: [push]
env:
  REPO: ${{ github.event.repository.name }}

jobs:
  #=========#
  # 1. Lint #
  #=========#
  lint:
    if: "! contains(toJSON(github.event.commits.*.msg), '[skip-ci]')"
    runs-on: ubuntu-18.04
    strategy:
      matrix:
        lang: ['php:7.4.1', 'php:5.6.40', 'php:5.2.17', 'js:12.14.0']

    container: elegantthemes/gh-action-lint-${{ matrix.lang }}
    steps:
      #======================#
      # 1.1 Initialize Repos #
      #======================#
      - name: Maybe Change REPO Environment Variable
        if: contains(github.event.repository.name, 'submodule')
        shell: bash
        run: echo "::set-env name=REPO::Divi"

      - name: Checkout Meta Repo
        uses: actions/checkout@v2
        with:
          repository: elegantthemes/meta
          ref: master
          path: workspace/meta
          token: ${{ secrets.ETSTAGING_TOKEN }}

      - name: Init Submodules &amp; Checkout Branches
        working-directory: workspace/meta
        shell: bash
        run: |
          git config --global http.extraheader "$(git config --get http.https://github.com/.extraheader)"
          git config --global protocol.version 2
          git submodule update --init --force --depth=1 --remote core-products
          cd core-products
          git submodule update --init --force --depth=1 --remote --recursive ${REPO}
          cd "${REPO}"
          git checkout "${GITHUB_REF}" || true
          git submodule foreach --recursive "git -c \"http.extraheader=$auth\" -c protocol.version=2 checkout \"${GITHUB_REF}\" || true"
</code></pre>
<p><a href="https://drive.google.com/file/d/1DKAFecf0FPKXBBakNobdC6m6Ln7VEH-a/view?usp=sharing" rel="nofollow noopener">Here</a> are the logs from the failed workflow run. Any idea what the problem is?</p>
<p><strong>Update: </strong></p>
<p>It looks like the only environment variables applied to the container are _ <strong>GITHUB_ACTIONS</strong> _ and _ <strong>HOME</strong> <em>.  AFAICT, </em> <strong>GITHUB_WORKSPACE</strong> _ is not applied when the runner starts the container. Not sure if this is even related to the issue though.</p>
<p><strong>Update 2:</strong></p>
<p>My image has git 2.17.1 but it looks like the checkout action requires git 2.18+. Rebuilding my images with latest git  now to see if that resolves the issue.</p>

**Comment by BrightRan** [comment, marked as answer]
<p>As your mentioned, you setup Git 2.17.1 in your Docker image, not Git 2.18 or higher. When executing checkout step, if Git 2.18 or higher is not in the PATH, the <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Factions%2Fcheckout%23whats-new&amp;data=02%7C01%7Cv-brira%40microsoft.com%7C0066c24a3a33446a46d008d79809a09b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145039428815237&amp;sdata=wqaGgUXXOvRPhLknBak7WTBZujKvK01EMJtqzNXJ%2BNI%3D&amp;reserved=0" rel="nofollow noopener">checkout action</a> will fallback to use the GitHub REST API to download the files. The API only can download the files as a ZIP with specified commit version and then extract the ZIP, it will not create a local branch. So when executing git command in the directory where the downloaded files are located, it will return the error “ <strong>not a git repository</strong> ”.</p>
<p>There are 2 workarounds as references:</p>
<ol>
<li>Directly execute the git command to checkout the repository and it submodules. Reference here: <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgit-scm.com%2Fbook%2Fen%2Fv2%2FGit-Tools-Submodules&amp;data=02%7C01%7Cv-brira%40microsoft.com%7C0066c24a3a33446a46d008d79809a09b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145039428820235&amp;sdata=CmMlOLhHJvAhxzDyK%2Fb7qHGZf%2BxFAoh3RZCPwtG1bIo%3D&amp;reserved=0" rel="nofollow noopener">https://git-scm.com/book/en/v2/Git-Tools-Submodules</a>
</li>
<li>Setup Git 2.18 or higher in the Docker image, and also need to set Git to the container’s PATH.</li>
</ol>
